src/manifold/src/csg_tree.cpp

```diff
@@ -340,14 +336,76 @@ void CsgOpNode::BatchBoolean(
  * `BatchBoolean` instead of using `Compose` for non-intersecting manifolds.
  */
 void CsgOpNode::BatchUnion() const {
-  std::vector<std::shared_ptr<const Manifold::Impl>> impls;
+  // INVARIANT: children_ is a vector of leaf nodes
+  // this kMaxUnionSize is a heuristic to avoid the pairwise disjoint check
+  // with O(n^2) complexity to take too long.
+  // If the number of children exceeded this limit, we will operate on chunks
+  // with size kMaxUnionSize.
+  constexpr int kMaxUnionSize = 1000;
   auto &children_ = impl_->children_;
-  for (auto &child : children_) {
-    impls.push_back(std::dynamic_pointer_cast<CsgLeafNode>(child)->GetImpl());
+  while (children_.size() > 1) {
+    int start;
+    if (children_.size() > kMaxUnionSize) {
+      start = children_.size() - kMaxUnionSize;
+    } else {
+      start = 0;
+    }
+    VecDH<Box> boxes;
+    boxes.reserve(children_.size() - start);
+    for (int i = start; i < children_.size(); i++) {
+      boxes.push_back(std::dynamic_pointer_cast<CsgLeafNode>(children_[i])
+                          ->GetImpl()
+                          ->bBox_);
+      const float precision =
+          std::dynamic_pointer_cast<CsgLeafNode>(children_[i])
+              ->GetImpl()
+              ->precision_;
+      boxes.back().max += glm::vec3(precision);
+      boxes.back().min -= glm::vec3(precision);
+    }
+    const Box *boxesD = boxes.cptrD();
+    // partition the children into a set of disjoint sets
+    // each set contains a set of children that are pairwise disjoint
+    std::vector<VecDH<size_t>> disjointSets;
+    for (size_t i = 0; i < boxes.size(); i++) {
+      auto lambda = [boxesD, i](const VecDH<size_t> &set) {
```

Comment:
I think CUDA doesn't like your lambda; should work if you make it a functor instead. Or possibly just add `__host__ __device__`?

Reply:
Yeah I forgot to add the `__host__ __device__` again.